src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/HtmlConventions.cs

```diff
@@ -52,6 +40,47 @@ internal static bool IsInvalidNonWhitespaceHtmlCharacters(char testChar)
     /// </example>
     public static string ToHtmlCase(string name)
     {
-        return HtmlCaseRegex.Replace(name, HtmlCaseRegexReplacement).ToLowerInvariant();
+        if (string.IsNullOrEmpty(name))
+        {
+            return name;
+        }
+
+        var input = name.AsSpan();
+
+        using var _ = StringBuilderPool.GetPooledObject(out var result);
+        result.SetCapacityIfLarger(input.Length + 5);
```

Comment:
>SetCapacityIfLarger [](http://example.com/codeflow?start=15&length=19)

I'm not a big fan of doing this on a pooled collection. Say you get a builder of length 5 and need one of length 6. You would then lose the exponential growth benefits of just letting the builder resize itself.

Reply:
I didn't think StringBuilder used exponential growth? At least on .NET Core I thought it used essentially a linked list of chunks. But I've no idea, and am more than willing to take your vast experience at looking at traces here.

Should I not pool? Or just not set the capacity? The benchmark certainly showed benefits from setting it, but I didn't do anything interesting like use a random input for it or anything, so very unrealistic.

Reply:
It's late for me, so I'm not sure how to read the StringBuilder code.

[https://github.com/dotnet/runtime/blob/2db9fdf5b4083c7459fc2150e8589a81ac7ebb39/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs#L282](https://github.com/dotnet/runtime/blob/2db9fdf5b4083c7459fc2150e8589a81ac7ebb39/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs#L282)

I'd vote to just simplify it and not set the capacity of pooled objects, they will grow to their proper size quickly enough, and by not setting the capacity, we ensure they will experience exponential growth doing so.

Reply:
Very interesting. Does it change if you use EnsureCapacity, and only change it if it increases?

Reply:
I guess I don't understand how enlarging capacity can be worse for performance. If you're lucky, it will never resize because it's big enough, and if you're unlucky it resizes once.

Without explicitly expanding, if you're lucky it's the same, and if you're unlucky it's at least 1 resize.

I'd love to learn what I'm missing here :)

I could see an argument for it being worse if it results in a pool full of huge mostly-empty string builders, but to me the answer would be to not pool.